---
title: "Single-cell RNA-Seq analysis of CD8 PD-1 Neg or Pos cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(sctransform)
library(SeuratDisk)
library(SeuratData)
library(plyr)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
```

# Comparison between CD8+ PD-1 negative and PD-1 positive cells by patient
Exploratory analysis of CD8+ PD-1 Neg and Pos T cells.
Follow guided clustering tutorial https://satijalab.org/seurat/v3.2/neg_5393k_tutorial.html and this one https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html, and https://satijalab.org/seurat/v3.2/integration.html (standard workflow)

## Patient 539

### Load Cell Ranger output and create separate Seurat object for positive and negative
```{r loading}
# Load dataset
neg_539.data <- Read10X(data.dir = "/t1-data/user/lfelce/scRNA-Seq/PD-1_CD8_T-cells/539_PD1_Neg_filtered_feature_bc_matrix")

# Initialize the Seurat object with gene expression data not antibody capture
neg_539 <- CreateSeuratObject(counts = neg_539.data$`Gene Expression`, project = "Patient 539", min.cells = 3, min.features = 200)

# Load dataset
pos_539.data <- Read10X(data.dir = "/t1-data/user/lfelce/scRNA-Seq/PD-1_CD8_T-cells/539_PD1_Pos_filtered_feature_bc_matrix")

# Initialize the Seurat object with gene expression data not antibody capture
pos_539 <- CreateSeuratObject(counts = pos_539.data$`Gene Expression`, project = "Patient 539", min.cells = 3, min.features = 200)

```

### Pre-processing and normalisation
```{r preprocessing}
# PD1 Neg
neg_539[["percent.mt"]] <- PercentageFeatureSet(neg_539, pattern = "^MT-")

FeatureScatter(neg_539, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

neg_539 <- subset(neg_539, subset = nFeature_RNA > 200 & nFeature_RNA < 4500 & percent.mt < 5)

neg_539 <- NormalizeData(neg_539, normalization.method = "LogNormalize", scale.factor = 10000)

# PD1 Pos
pos_539[["percent.mt"]] <- PercentageFeatureSet(pos_539, pattern = "^MT-")

FeatureScatter(pos_539, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

pos_539 <- subset(pos_539, subset = nFeature_RNA > 200 & nFeature_RNA < 4500 & percent.mt < 5)

pos_539 <- NormalizeData(pos_539, normalization.method = "LogNormalize", scale.factor = 10000)

```

### Find variable features and scale data
```{r variable features}
neg_539 <- FindVariableFeatures(neg_539, selection.method = "vst", nfeatures = 2000)
pos_539 <- FindVariableFeatures(pos_539, selection.method = "vst", nfeatures = 2000)

neg_539 <- ScaleData(object = neg_539, vars.to.regress = c("nCount_RNA", "percent.mt"))
pos_539 <- ScaleData(object = pos_539, vars.to.regress = c("nCount_RNA", "percent.mt"))

```

### Run PCA

```{r pca}
neg_539 <- RunPCA(object = neg_539,  npcs = 30, verbose = FALSE)

# standard PCA plot
DimPlot(object = neg_539, reduction = "pca")

pos_539 <- RunPCA(object = pos_539,  npcs = 30, verbose = FALSE)

# standard PCA plot
DimPlot(object = pos_539, reduction = "pca")
```

### Determine statistically significant principal components

```{r jackstraw, fig.height=4, fig.width=6}
neg_539 <- JackStraw(neg_539, num.replicate = 100)
neg_539 <- ScoreJackStraw(neg_539, dims = 1:20)
JackStrawPlot(neg_539, dims = 1:20)

pos_539 <- JackStraw(pos_539, num.replicate = 100)
pos_539 <- ScoreJackStraw(pos_539, dims = 1:20)
JackStrawPlot(pos_539, dims = 1:20)

```

### Cell clustering and UMAP

```{r clustering, fig.height=6, fig.width=16}
# calculate KNN and construct SNN graph (find neighbours), then find clusters
neg_539 <- FindNeighbors(neg_539, dims = 1:20)
neg_539 <- FindClusters(neg_539, resolution = 0.4)
neg_539 <- RunUMAP(neg_539, dims = 1:20)
UMAP_neg_cluster <- DimPlot(neg_539, reduction = "umap", group.by = "seurat_clusters") + ggtitle("Patient 539 PD-1 Neg") 

pos_539 <- FindNeighbors(pos_539, dims = 1:20)
pos_539 <- FindClusters(pos_539, resolution = 0.4)
pos_539 <- RunUMAP(pos_539, dims = 1:20)
UMAP_pos_cluster <- DimPlot(pos_539, reduction = "umap", group.by = "seurat_clusters")  + ggtitle("Patient 539 PD-1 Pos")

UMAP_neg_cluster + UMAP_pos_cluster
```
### Find All Markers
```{r}
setwd('/t1-data/user/lfelce/scRNA-Seq/PD-1_CD8_T-cells/')

neg_539.markers <- FindAllMarkers(object = neg_539, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
write.csv (neg_539.markers, "neg_539_markers.csv")

pos_539.markers <- FindAllMarkers(object = pos_539, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25)
write.csv (pos_539.markers, "pos_539_markers.csv")

```

### Use Calli's list for cell annotation

```{r calli's markers}
tmarkers <- read.csv("Markers_for_Ling.csv", header=F)

# PD1 Neg
cluster_0 <- neg_539.markers[neg_539.markers$cluster == "0", ]
cluster_1 <- neg_539.markers[neg_539.markers$cluster == "1", ]
cluster_2 <- neg_539.markers[neg_539.markers$cluster == "2", ]
cluster_3 <- neg_539.markers[neg_539.markers$cluster == "3", ]
cluster_4 <- neg_539.markers[neg_539.markers$cluster == "4", ]
cluster_5 <- neg_539.markers[neg_539.markers$cluster == "5", ]
cluster_6 <- neg_539.markers[neg_539.markers$cluster == "6", ]
cluster_7 <- neg_539.markers[neg_539.markers$cluster == "7", ]
cluster_8 <- neg_539.markers[neg_539.markers$cluster == "8", ]

cluster_0 <- cluster_0[cluster_0$p_val_adj < 0.05,]
cluster_1 <- cluster_1[cluster_1$p_val_adj < 0.05,]
cluster_2 <- cluster_2[cluster_2$p_val_adj < 0.05,]
cluster_3 <- cluster_3[cluster_3$p_val_adj < 0.05,]
cluster_4 <- cluster_4[cluster_4$p_val_adj < 0.05,]
cluster_5 <- cluster_5[cluster_5$p_val_adj < 0.05,]
cluster_6 <- cluster_6[cluster_6$p_val_adj < 0.05,]
cluster_7 <- cluster_7[cluster_7$p_val_adj < 0.05,]
cluster_8 <- cluster_8[cluster_8$p_val_adj < 0.05,]


genes_0 <- tmarkers[is.element(tmarkers$V1, cluster_0$gene),]
genes_1 <- tmarkers[is.element(tmarkers$V1, cluster_1$gene),]
genes_2 <- tmarkers[is.element(tmarkers$V1, cluster_2$gene),]
genes_3 <- tmarkers[is.element(tmarkers$V1, cluster_3$gene),]
genes_4 <- tmarkers[is.element(tmarkers$V1, cluster_4$gene),]
genes_5 <- tmarkers[is.element(tmarkers$V1, cluster_5$gene),]
genes_6 <- tmarkers[is.element(tmarkers$V1, cluster_6$gene),]
genes_7 <- tmarkers[is.element(tmarkers$V1, cluster_7$gene),]
genes_8 <- tmarkers[is.element(tmarkers$V1, cluster_8$gene),]

# PD1 Pos

```





### Find anchors and integrate

```{r integrate}

pancreas.anchors <- FindIntegrationAnchors(object.list = c(neg_539, pos_539), dims = 1:30)

anchors.539 <- pancreas.anchors

integrated.539 <- IntegrateData(anchorset = anchors.539, dims = 1:30)

DefaultAssay(integrated.539) <- "integrated"

# Run the standard workflow for visualization and clustering
integrated.539 <- ScaleData(integrated.539, verbose = FALSE)
integrated.539 <- RunPCA(integrated.539, npcs = 30, verbose = FALSE)
integrated.539 <- RunUMAP(integrated.539, reduction = "pca", dims = 1:30)
p1 <- DimPlot(integrated.539, reduction = "umap", group.by = "tech")
p2 <- DimPlot(integrated.539, reduction = "umap", group.by = "celltype", label = TRUE, 
    repel = TRUE) + NoLegend()
p1 + p2

```


### Save

```{r save}
saveRDS(neg_539, file = "neg_539_seurat.rds")
saveRDS(pos_539, file = "pos_539_seurat.rds")
```


