---
title: "Single-cell RNA-Seq analysis of CD8+ SARS-CoV-2 specific T cells"
output:
  word_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Seurat)
library(sctransform)
library(SeuratDisk)
library(SeuratData)
library(plyr)
library(dplyr)
library(patchwork)
library(cowplot)
library(ggplot2)
library(data.table)
library(tidyverse)
```

# Comparison between CD8+ PD-1 negative and PD-1 positive cells by patient
Exploratory analysis of CD8+ PD-1 Neg and Pos T cells.
Follow guided clustering tutorial https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html and this one https://scrnaseq-course.cog.sanger.ac.uk/website/seurat-chapter.html, and https://satijalab.org/seurat/v3.2/integration.html (standard workflow)

## Patient 539

### Load Cell Ranger output and create separate Seurat object for positive and negative
```{r loading}
# Load dataset
neg_539.data <- Read10X(data.dir = "/t1-data/user/lfelce/scRNA-Seq/PD-1_CD8_T-cells/539_PD1_Neg_filtered_feature_bc_matrix")

# Initialize the Seurat object with the raw (non-normalized data).
neg_539 <- CreateSeuratObject(counts = neg_539.data$`Gene Expression`, project = "Patient 539", min.cells = 3, min.features = 200)

# Load dataset
pos_539.data <- Read10X(data.dir = "/t1-data/user/lfelce/scRNA-Seq/PD-1_CD8_T-cells/539_PD1_Pos_filtered_feature_bc_matrix")

# Initialize the Seurat object with the raw (non-normalized data).
pos_539 <- CreateSeuratObject(counts = pos_539.data$`Gene Expression`, project = "Patient 539", min.cells = 3, min.features = 200)



```
